### PR TITLE
cmd/libsnap: fix parsing of empty mountinfo fields

### PR DESCRIPTION
**Pull request by zyga**
This patch fixes an apparently unexpected kernel behavior whereby
a tmpfs with empty "mount source" (which is customarily set to "none"
by various tools) results in a mountinfo field with two consecutive
spaces. That is, empty fields are just printed as-is.

The parser is adjusted and a new unit test is added to cover the case
where this was discovered.

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

File: cmd/libsnap-confine-private/mountinfo.c [hunk 1]
```diff
@@ -187,11 +187,20 @@ static struct sc_mountinfo_entry *sc_parse_mountinfo_entry(const char *line)
...
+			// Special case for empty fields which cannot be parsed with %s.
+			*field = '\0';
+			offset += 1;
+		} else {
+			int nscanned = sscanf(line + offset, "%s%n", field,
+					      &offset_delta);
+			if (nscanned != 1)
+				return NULL;
+			offset += offset_delta;
+			if (line[offset] == ' ') {
```
Comment (stolowski) [inline cmd/libsnap-confine-private/mountinfo.c]: I'm unclear why is this needed again here? Can you explain?

Comment (zyga) [reply cmd/libsnap-confine-private/mountinfo.c]: The test checks if the first character of the field is a space. This is only true if the field is empty and so this gets special-cased as `scanf` style functions cannot parse empty strings with `%s` format character.

Comment (stolowski) [reply cmd/libsnap-confine-private/mountinfo.c]: Ok, sorry, I should be more clear: my question is whether this is needed here if same test is done on line 190?

Comment (zyga) [reply cmd/libsnap-confine-private/mountinfo.c]: Ah, sorry. I looked at the wrong part of the diff.

This test is to compensate the change that was done to `sscanf` call. Earlier we parsed `%s %n` and now we parse `%s%n`. This makes us scan each field correctly. For a visual look at how this behaves:

```
zyga@fyke:~/snapd/cmd$ make -j libsnap-confine-private/unit-tests && ./libsnap-confine-private/unit-tests -p /mountinfo/parse_mountinfo_entry/empty_source
gcc -DHAVE_CONFIG_H -I.    -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -D_ENABLE_FAULT_INJECTION -g -O2 -MT libsnap-confine-private/libsnap_confine_private_unit_tests-mountinfo-test.o -MD -MP -MF libsnap-confine-private/.deps/libsnap_confine_private_unit_tests-mountinfo-test.Tpo -c -o libsnap-confine-private/libsnap_confine_private_unit_tests-mountinfo-test.o `test -f 'libsnap-confine-private/mountinfo-test.c' || echo './'`libsnap-confine-private/mountinfo-test.c
gcc -DHAVE_CONFIG_H -I.    -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -D_ENABLE_FAULT_INJECTION -g -O2 -MT libsnap-confine-private/libsnap_confine_private_unit_tests-utils-test.o -MD -MP -MF libsnap-confine-private/.deps/libsnap_confine_private_unit_tests-utils-test.Tpo -c -o libsnap-confine-private/libsnap_confine_private_unit_tests-utils-test.o `test -f 'libsnap-confine-private/utils-test.c' || echo './'`libsnap-confine-private/utils-test.c
mv -f libsnap-confine-private/.deps/libsnap_confine_private_unit_tests-utils-test.Tpo libsnap-confine-private/.deps/libsnap_confine_private_unit_tests-utils-test.Po
mv -f libsnap-confine-private/.deps/libsnap_confine_private_unit_tests-mountinfo-test.Tpo libsnap-confine-private/.deps/libsnap_confine_private_unit_tests-mountinfo-test.Po
gcc -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -D_ENABLE_FAULT_INJECTION -g -O2   -o libsnap-confine-private/unit-tests libsnap-confine-private/libsnap_confine_private_unit_tests-classic-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-cleanup-funcs-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-error-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-fault-injection-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-locking-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-mount-opt-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-mountinfo-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-privs-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-secure-getenv-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-snap-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-string-utils-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-test-utils.o libsnap-confine-private/libsnap_confine_private_unit_tests-test-utils-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-unit-tests-main.o libsnap-confine-private/libsnap_confine_private_unit_tests-unit-tests.o libsnap-confine-private/libsnap_confine_private_unit_tests-utils-test.o -lglib-2.0  -Wl,-Bstatic -lcap -Wl,-Bdynamic
/mountinfo/parse_mountinfo_entry/empty_source: Input buffer (first), with offset arrow
Output buffer (second)
 ------------v
>304 301 0:45 / /snap/test-snapd-content-advanced-plug/x1 rw,relatime - tmpfs  rw<
>################################################################################<
>================================================================================<
Input buffer (first), with offset arrow
Output buffer (second)
 --------------v
>304 301 0:45 / /snap/test-snapd-content-advanced-plug/x1 rw,relatime - tmpfs  rw<
>#############/@#################################################################<
>================================================================================<
Input buffer (first), with offset arrow
Output buffer (second)
 --------------------------------------------------------v
>304 301 0:45 / /snap/test-snapd-content-advanced-plug/x1 rw,relatime - tmpfs  rw<
>#############/@/snap/test-snapd-content-advanced-plug/x1@#######################<
>================================================================================<
Input buffer (first), with offset arrow
Output buffer (second)
 --------------------------------------------------------------------v
>304 301 0:45 / /snap/test-snapd-content-advanced-plug/x1 rw,relatime - tmpfs  rw<
>#############/@/snap/test-snapd-content-advanced-plug/x1@rw,relatime@###########<
>================================================================================<
Input buffer (first), with offset arrow
Output buffer (second)
 ----------------------------------------------------------------------v
>304 301 0:45 / /snap/test-snapd-content-advanced-plug/x1 rw,relatime - tmpfs  rw<
>#############/@/snap/test-snapd-content-advanced-plug/x1@rw,relatime@-@#########<
>================================================================================<
Input buffer (first), with offset arrow
Output buffer (second)
 ----------------------------------------------------------------------------v
>304 301 0:45 / /snap/test-snapd-content-advanced-plug/x1 rw,relatime - tmpfs  rw<
>#############/@/snap/test-snapd-content-advanced-plug/x1@rw,relatime@@@tmpfs@###<
>================================================================================<
Input buffer (first), with offset arrow
Output buffer (second)
 -----------------------------------------------------------------------------v
>304 301 0:45 / /snap/test-snapd-content-advanced-plug/x1 rw,relatime - tmpfs  rw<
>#############/@/snap/test-snapd-content-advanced-plug/x1@rw,relatime@@@tmpfs@@##<
>================================================================================<
Input buffer (first), with offset arrow
Output buffer (second)
 -------------------------------------------------------------------------------v
>304 301 0:45 / /snap/test-snapd-content-advanced-plug/x1 rw,relatime - tmpfs  rw<
>#############/@/snap/test-snapd-content-advanced-plug/x1@rw,relatime@@@tmpfs@@rw<
>================================================================================<
Input buffer (first), with offset arrow
Output buffer (second)
 -------------------------------------------------------------------------------v
>304 301 0:45 / /snap/test-snapd-content-advanced-plug/x1 rw,relatime - tmpfs  rw<
>#############/@/snap/test-snapd-content-advanced-plug/x1@rw,relatime@@@tmpfs@@rw<
>================================================================================<
```

File: cmd/libsnap-confine-private/mountinfo-test.c [hunk 2]
```diff
@@ -156,6 +156,28 @@ static void test_parse_mountinfo_entry__many_tags()
 	g_assert_null(entry->next);
 }
 
+static void test_parse_mountinfo_entry__empty_source()
+{
+	const char *line =
+	    "304 301 0:45 / /snap/test-snapd-content-advanced-plug/x1 rw,relatime - tmpfs  rw";
```
Comment (jdstrand) [inline cmd/libsnap-confine-private/mountinfo-test.c]: The two spaces between 'tmpfs' and 'rw': is this _guaranteed_ to denote an empty source? Ie, is it an implementation artifact or is it part of the specification for mountinfo? This seems scarily brittle at first glance.

Comment (zyga) [reply cmd/libsnap-confine-private/mountinfo-test.c]: It is just how the kerne is impemented. Reading the specification it is neither violating it nor explicitly explained that it behaves this way. IMO this is a kernel bug but the way countess kernel interfaces are defined it doesn’t stand out. I for one would love to see standardized serializer but this is unlikely to happen anytime soon.

I looked at how this is implemented and it is a %s with some escape logic for spaces and other characters.
